core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java

```diff
@@ -105,20 +110,51 @@ public void testLocalNodeIsMaster() {
 
     /**
      * Test that the indices created and indices deleted lists between two cluster states
-     * are correct when there is no change in the cluster UUID.  Also tests metadata equality
-     * between cluster states.
+     * are correct when there is a change in indices added and deleted.  Also tests metadata
+     * equality between cluster states.
      */
-    public void testMetaDataChangesOnNoMasterChange() {
-        metaDataChangesCheck(false);
-    }
-
-    /**
-     * Test that the indices created and indices deleted lists between two cluster states
-     * are correct when there is a change in the cluster UUID.  Also tests metadata equality
-     * between cluster states.
-     */
-    public void testMetaDataChangesOnNewClusterUUID() {
-        metaDataChangesCheck(true);
+    public void testIndicesMetaDataChanges() {
+        final int numNodesInCluster = 3;
+        for (int i = 0; i < INDICES_CHANGE_NUM_TESTS; i++) {
+            final ClusterState previousState = createState(numNodesInCluster, randomBoolean(), initialIndices);
+            final int numAdd = randomIntBetween(0, 5); // add random # of indices to the next cluster state
+            int numDel = randomIntBetween(0, initialIndices.size() - 1); // delete random # of indices from the next cluster state
+            final List<Index> addedIndices = addIndices(numAdd);
+            List<Index> delIndices = delIndices(numDel, initialIndices);
+            final ClusterState newState = nextState(previousState, addedIndices, delIndices, 0);
+            ClusterChangedEvent event = new ClusterChangedEvent("_na_", newState, previousState);
+            final List<String> addsFromEvent = event.indicesCreated();
+            List<Index> delsFromEvent = event.indicesDeleted();
+            Collections.sort(addsFromEvent);
+            Collections.sort(delsFromEvent, SORT_FUNCTION);
+            assertThat(addsFromEvent, equalTo(addedIndices.stream().map((s) -> s.getName()).collect(Collectors.toList())));
+            assertThat(delsFromEvent, equalTo(delIndices));
+            assertThat(event.metaDataChanged(), equalTo(addedIndices.size() > 0 || delIndices.size() > 0));
+            final IndexGraveyard newGraveyard = event.state().metaData().indexGraveyard();
+            final IndexGraveyard oldGraveyard = event.previousState().metaData().indexGraveyard();
+            assertThat(newGraveyard.diff(oldGraveyard).getAdded().size(), equalTo(delIndices.size()));
+
+            // test a second cluster state update with more index deletions to test the tombstone diffs
+            addedIndices.clear();
```

Comment:
can't we incorporate this into the loop? keep the old cluster state around and keep changing it?